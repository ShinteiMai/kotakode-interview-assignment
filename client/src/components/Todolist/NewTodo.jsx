import React from "react";
import PropTypes from "prop-types";
import { Formik, Field } from "formik";
import { Button, Box, Input } from "@chakra-ui/core";
import {
  submitTodoButtonComponent,
  addTodoFormComponent,
} from "../../utils/testConstants";

const NewTodo = ({ createTodo, setIsAddingTodo }) => (
  <Formik
    initialValues={{ task: "" }}
    onSubmit={({ task }) => {
      createTodo(task);
      setIsAddingTodo(false);
    }}
  >
    {({ handleSubmit }) => (
      <form onSubmit={handleSubmit} data-test={addTodoFormComponent}>
        <Box display="flex" alignItems="top" mx="auto" justifyContent="center">
          <Box>
            <Field
              placeholder="e.g. Wash dishes"
              type="text"
              name="task"
              as={Input}
            />
            <Button type="submit">Add task</Button>
          </Box>
          <Button
            data-test={submitTodoButtonComponent}
            onClick={() => setIsAddingTodo(false)}
          >
            Cancel
          </Button>
        </Box>
      </form>
    )}
  </Formik>
);

NewTodo.propTypes = {
  createTodo: PropTypes.func,
  setIsAddingTodo: PropTypes.func,
};

export default NewTodo;
