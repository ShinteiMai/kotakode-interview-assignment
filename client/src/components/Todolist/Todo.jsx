import React, { useState } from "react";
import PropTypes from "prop-types";
import { Formik, Field } from "formik";
import { ListItem } from "@chakra-ui/core";

export const taskPlaceholderValue = {
  unfocused: "Click here to add a task!",
  focused: "Add a task!",
};

const Todo = ({ mutateTodos, task, index }) => {
  const [placeholder, setPlaceholder] = useState(
    taskPlaceholderValue.unfocused
  );

  return (
    <ListItem>
      <Formik
        initialValues={{
          task,
        }}
        onSubmit={({ task }) => {
          mutateTodos(index, task);
        }}
      >
        {({ handleSubmit }) => (
          <form onSubmit={handleSubmit}>
            <Field
              type="text"
              name="task"
              placeholder={placeholder}
              onFocus={() => {
                setPlaceholder(taskPlaceholderValue.focused);
              }}
              onBlur={() => {
                setPlaceholder(taskPlaceholderValue.unfocused);
              }}
            />
          </form>
        )}
      </Formik>
    </ListItem>
  );
};

Todo.propTypes = {
  mutateTodos: PropTypes.func,
  task: PropTypes.string,
  index: PropTypes.number,
};

export default Todo;
