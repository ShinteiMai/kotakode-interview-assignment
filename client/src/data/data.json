{
  "tasks": ["Wash the dishes", "Cook a meal"]
}
