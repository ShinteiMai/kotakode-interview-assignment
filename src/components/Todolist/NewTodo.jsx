import React, { useState } from "react";
import PropTypes from "prop-types";
import { Formik, Field } from "formik";
import { Button, Box, Input, ButtonGroup } from "@chakra-ui/core";
import { AiFillSave } from "react-icons/ai";
import { TiCancel } from "react-icons/ti";
import Dropdown from "react-dropdown";
import {
  submitTodoButtonComponent,
  addTodoFormComponent,
} from "../../utils/testConstants";

import "react-dropdown/style.css";
import Flags from "../Flags";
import { todoSchema } from "../../schema/taskSchema";

const priorityFlags = Flags();
const NewTodo = ({ createTodo, setIsAddingTodo }) => {
  const [isPriorityTouched, setIsPriorityTouched] = useState(false);

  return (
    <Formik
      initialValues={{ task: "", priority: "" }}
      onSubmit={({ task, priority }) => {
        createTodo(task, priority);
        setIsAddingTodo(false);
      }}
      validationSchema={todoSchema}
    >
      {({ handleSubmit, errors, touched, setFieldValue, values }) => (
        <form onSubmit={handleSubmit} data-test={addTodoFormComponent}>
          <Box
            display="flex"
            flexDir="column"
            width="100%"
            alignItems="top"
            justifyContent="center"
          >
            <Box width={["60%", "50%", "40%"]}>
              <Box mb={4}>
                <Field
                  placeholder="e.g. Wash dishes"
                  type="text"
                  name="task"
                  as={Input}
                />
                {errors.task && touched.task ? (
                  <Box
                    display="inline-block"
                    fontWeight="500"
                    color="red.500"
                    ml={4}
                    mt={2}
                  >
                    {errors.task}
                  </Box>
                ) : null}
              </Box>
              <Box mb={4}>
                <Dropdown
                  options={[1, 2, 3, 4]}
                  onChange={(option) => {
                    setFieldValue("priority", option.value);
                    setIsPriorityTouched(true);
                  }}
                  placeholder={
                    isPriorityTouched
                      ? priorityFlags[values.priority - 1]
                      : "Priority"
                  }
                />
                {errors.priority && touched.priority ? (
                  <Box
                    ml={4}
                    mt={2}
                    fontWeight="500"
                    display="inline-block"
                    color="red.500"
                  >
                    {errors.priority}
                  </Box>
                ) : null}
              </Box>
            </Box>
            <ButtonGroup>
              <Button
                type="submit"
                leftIcon={AiFillSave}
                variantColor="yellow"
                variant="outline"
              >
                Add task
              </Button>
              <Button
                data-test={submitTodoButtonComponent}
                leftIcon={TiCancel}
                variantColor="red"
                variant="outline"
                onClick={() => setIsAddingTodo(false)}
              >
                Cancel
              </Button>
            </ButtonGroup>
          </Box>
        </form>
      )}
    </Formik>
  );
};

NewTodo.propTypes = {
  createTodo: PropTypes.func.isRequired,
  setIsAddingTodo: PropTypes.func.isRequired,
};

export default NewTodo;
