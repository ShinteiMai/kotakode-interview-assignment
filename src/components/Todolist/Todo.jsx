import React, { useState } from "react";
import PropTypes from "prop-types";
import { Formik, Field } from "formik";
import Dropdown from "react-dropdown";
import { ListItem, Input, Box, Text } from "@chakra-ui/core";
import { AiFillEdit, AiFillSave } from "react-icons/ai";
import { FaRegCircle } from "react-icons/fa";
import Flags from "../Flags";

import "react-dropdown/style.css";
import { todoSchema } from "../../schema/taskSchema";
import { todosComponent } from "../../utils/testConstants";

const colors = ["red.500", "orange.300", "yellow.300", "green.400"];
export const taskPlaceholderValue = {
  unfocused: "Click here to add a task!",
  focused: "Add a task!",
};

const priorityFlags = Flags();

const Todo = ({ updateTodo, deleteTodo, task, index }) => {
  const [isEditing, setIsEditing] = useState(false);
  const [isPriorityTouched, setIsPriorityTouched] = useState(false);
  const [placeholder, setPlaceholder] = useState(
    taskPlaceholderValue.unfocused
  );

  return (
    <ListItem data-testid={todosComponent}>
      <Formik
        initialValues={{
          task: task.task,
          priority: task.priority,
        }}
        onSubmit={({ task, priority }) => {
          updateTodo(index, task, priority);
        }}
        validationSchema={todoSchema}
      >
        {({
          handleSubmit,
          values,
          setFieldValue,
          errors,
          touched,
          submitForm,
        }) => (
          <form onSubmit={handleSubmit}>
            <Box
              display="flex"
              alignItems="center"
              flexDir={isEditing ? ["column", "row"] : "row"}
            >
              <Box
                display="flex"
                alignItems="center"
                w={["100%", "60%", "50%"]}
                mb={[3, 0]}
              >
                <Box>
                  <Box
                    style={{
                      cursor: "pointer",
                    }}
                    mr={2}
                    mb={1}
                    as={FaRegCircle}
                    color={colors[values.priority - 1]}
                    onClick={() => deleteTodo(index)}
                  />
                </Box>
                <Box display="flex" alignItems="center" mr={4} w="80%">
                  {isEditing ? (
                    <Box>
                      <Field
                        type="text"
                        name="task"
                        as={Input}
                        placeholder={placeholder}
                        onFocus={() => {
                          setPlaceholder(taskPlaceholderValue.focused);
                        }}
                        onBlur={() => {
                          setPlaceholder(taskPlaceholderValue.unfocused);
                        }}
                      />
                    </Box>
                  ) : (
                    <Box>
                      <Text>{values.task}</Text>
                    </Box>
                  )}
                </Box>
              </Box>
              <Box display="flex">
                {isEditing ? (
                  <Box ml={4}>
                    <Dropdown
                      options={[1, 2, 3, 4]}
                      onChange={(option) => {
                        setFieldValue("priority", option.value);
                        setIsPriorityTouched(true);
                      }}
                      placeholder={
                        isPriorityTouched
                          ? priorityFlags[values.priority - 1]
                          : "Priority"
                      }
                    />
                  </Box>
                ) : null}
                <Box
                  style={{
                    cursor: "pointer",
                  }}
                  display="flex"
                  alignItems="center"
                  ml={4}
                >
                  <Box
                    size={5}
                    as={isEditing ? AiFillSave : AiFillEdit}
                    onClick={() => {
                      if (isEditing) submitForm();
                      if (!errors.task && !errors.priority) {
                        setIsEditing(!isEditing);
                      }
                    }}
                  />
                </Box>
              </Box>
            </Box>
            <Box ml={8}>
              {errors.task && touched.task ? (
                <Text color="red.500">{errors.task}</Text>
              ) : null}
              {errors.priority && touched.priority ? (
                <Text color="red.500">{errors.priority}</Text>
              ) : null}
            </Box>
          </form>
        )}
      </Formik>
    </ListItem>
  );
};

Todo.propTypes = {
  task: PropTypes.shape({
    id: PropTypes.string.isRequired,
    task: PropTypes.string.isRequired,
    priority: PropTypes.number.isRequired,
  }).isRequired,
  updateTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired,
};

export default Todo;
