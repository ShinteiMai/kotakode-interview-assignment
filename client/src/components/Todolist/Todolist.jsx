import React, { useState } from "react";
import PropTypes from "prop-types";

import { List, Button } from "@chakra-ui/core";

import Todo from "./Todo";

const Todolist = ({ tasks }) => {
  const [todos, setTodos] = useState(tasks);

  const mutateTodos = (index, task) => {
    const newTodos = [...todos];
    newTodos[index] = task;
    setTodos(newTodos);
  };

  console.log(todos);

  return (
    <>
      <div>A todolist!</div>
      <div>
        <List spacing={3}>
          {todos.map((todo, index) => (
            <Todo
              key={index}
              mutateTodos={mutateTodos}
              task={todo}
              index={index}
            />
          ))}
        </List>
      </div>
      <Button variantColor="teal" onClick={() => setTodos(todos.concat(""))}>
        Add a new task
      </Button>
    </>
  );
};

Todolist.propTypes = {
  tasks: PropTypes.arrayOf(PropTypes.string),
};

export default Todolist;
