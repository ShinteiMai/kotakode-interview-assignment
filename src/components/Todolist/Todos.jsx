import React from "react";
import PropTypes from "prop-types";

import { List, Box } from "@chakra-ui/core";

import Todo from "./Todo";

import {
  todosComponent,
  todosContainerComponent,
} from "../../utils/testConstants";
import { DragDropContext, Droppable, Draggable } from "react-beautiful-dnd";
import { reorder } from "../../utils/reorder";

const getItemStyle = (isDragging, draggableStyle) => ({
  userSelect: "none",
  ...draggableStyle,
});

const Todos = ({ todos, setTodos, updateTodo, deleteTodo }) => {
  return (
    <DragDropContext
      onDragEnd={({ source, destination }) => {
        if (!destination) return;

        setTodos(reorder(todos, source.index, destination.index));
      }}
    >
      <Droppable droppableId="droppable-1">
        {(provided) => (
          <List
            {...provided.droppableProps}
            ref={provided.innerRef}
            data-testid={todosContainerComponent}
          >
            {todos.map((todo, index) => (
              <Draggable key={todo.id} draggableId={todo.id} index={index}>
                {(provided, snapshot) => (
                  <Box
                    style={getItemStyle(
                      snapshot.isDragging,
                      provided.draggableProps.style
                    )}
                    ref={provided.innerRef}
                    {...provided.draggableProps}
                    {...provided.dragHandleProps}
                    mb={4}
                  >
                    <Todo
                      key={index}
                      updateTodo={updateTodo}
                      deleteTodo={deleteTodo}
                      task={todo}
                      index={index}
                    />
                  </Box>
                )}
              </Draggable>
            ))}
            {provided.placeholder}
          </List>
        )}
      </Droppable>
    </DragDropContext>
  );
};

Todos.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      task: PropTypes.string.isRequired,
      priority: PropTypes.number.isRequired,
    })
  ).isRequired,
  setTodos: PropTypes.func.isRequired,
  updateTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
};

export default Todos;
