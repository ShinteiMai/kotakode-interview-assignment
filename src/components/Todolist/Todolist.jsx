import React, { useState } from "react";
import PropTypes from "prop-types";

import moment from "moment";

import { Text, Button, Box, Divider } from "@chakra-ui/core";
import { AiOutlinePlus } from "react-icons/ai";

import NewTodo from "./NewTodo";
import Todos from "./Todos";

const Todolist = ({ tasks }) => {
  const [todos, setTodos] = useState(
    tasks.map((task, index) => {
      return {
        id: `item-${index}`,
        task: task.task,
        priority: task.priority,
      };
    })
  );
  const [isAddingTodo, setIsAddingTodo] = useState(false);

  const createTodo = (task, priority) =>
    setTodos(todos.concat({ id: `item-${todos.length + 1}`, task, priority }));

  const updateTodo = (index, task, priority) => {
    const newTodos = [...todos];
    newTodos[index] = { ...newTodos[index], task, priority };
    setTodos(newTodos);
  };

  const deleteTodo = (index) => {
    const newTodos = [...todos];
    newTodos.splice(index, 1);
    setTodos(newTodos);
  };

  return (
    <>
      <Box mb={6}>
        <Text fontWeight={600} fontSize={["1xl", "2xl"]}>
          {moment().format("dddd, MMMM Do YYYY")}
        </Text>
        {isAddingTodo ? (
          <NewTodo createTodo={createTodo} setIsAddingTodo={setIsAddingTodo} />
        ) : (
          <Button
            leftIcon={AiOutlinePlus}
            variantColor="yellow"
            variant="outline"
            onClick={() => setIsAddingTodo(true)}
          >
            Add task
          </Button>
        )}
      </Box>
      <Divider opacity={0.3} />
      <Box minHeight={64}>
        <Todos
          todos={todos}
          setTodos={setTodos}
          updateTodo={updateTodo}
          deleteTodo={deleteTodo}
        />
      </Box>
    </>
  );
};

Todolist.propTypes = {
  tasks: PropTypes.arrayOf(
    PropTypes.shape({
      task: PropTypes.string.isRequired,
      priority: PropTypes.number.isRequired,
    })
  ),
};

export default Todolist;
