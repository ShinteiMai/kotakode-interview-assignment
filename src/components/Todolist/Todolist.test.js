import React from "react";
import { GlobalStyle } from "@chakra-ui/system";
import { theme, ThemeProvider } from "@chakra-ui/core";
import { shallow } from "enzyme";
import { render, fireEvent, waitFor } from "@testing-library/react";

import Todolist from "./Todolist";

import { tasks } from "../../data/data.json";
import {
  todosComponent,
  addTodoButtonComponent,
  todosContainerComponent,
  newTodoComponent,
  taskFormComponent,
  submitTodoButtonComponent,
} from "../../utils/testConstants";

const defaultProps = { tasks };

describe("Todolist CRUD functionality", () => {
  it("should shallow render correctly", async () => {
    const wrapper = shallow(<Todolist {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should be able to render correctly", async () => {
    const { getByTestId } = render(
      <ThemeProvider value={theme}>
        <GlobalStyle />
        <Todolist {...defaultProps} />
      </ThemeProvider>
    );

    expect(getByTestId(addTodoButtonComponent)).toBeInTheDocument();
  });

  it("should able to render todos correctly", async () => {
    const { getByTestId, getAllByTestId } = render(
      <ThemeProvider value={theme}>
        <GlobalStyle />
        <Todolist {...defaultProps} />
      </ThemeProvider>
    );

    expect(getByTestId(todosContainerComponent)).toBeInTheDocument();
    expect(getAllByTestId(todosComponent)).toHaveLength(tasks.length);
  });

  it("should be able to add a new todo", async () => {
    const data = {
      task: "Workout",
      priority: 2,
    };

    const { getByTestId, queryByTestId, getAllByTestId } = render(
      <ThemeProvider value={theme}>
        <GlobalStyle />
        <Todolist {...defaultProps} />
      </ThemeProvider>
    );

    expect(queryByTestId(newTodoComponent)).not.toBeInTheDocument();

    expect(getByTestId(addTodoButtonComponent)).toBeInTheDocument();
    fireEvent.click(getByTestId(addTodoButtonComponent));
    await waitFor(() => getByTestId(newTodoComponent));

    expect(getByTestId(newTodoComponent)).toBeInTheDocument();

    expect(getByTestId(taskFormComponent)).toBeInTheDocument();
    const taskForm = getByTestId(taskFormComponent);
    fireEvent.change(taskForm, { target: { value: data.task } });

    expect(getByTestId(submitTodoButtonComponent)).toBeInTheDocument();
    fireEvent.click(getByTestId(submitTodoButtonComponent));

    await waitFor(() => getAllByTestId(todosComponent));
    expect(queryByTestId(newTodoComponent)).not.toBeInTheDocument();
    expect(queryByTestId(addTodoButtonComponent)).toBeInTheDocument();
    expect(getAllByTestId(todosComponent)).toHaveLength(tasks.length + 1);
  });

  it("should be abled to delete a todo", async () => {});

  it("should be able to update a todo", async () => {});
});
