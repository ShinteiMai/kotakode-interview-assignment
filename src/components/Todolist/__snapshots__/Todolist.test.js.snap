// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todolist CRUD functionality should shallow render correctly 1`] = `ShallowWrapper {}`;
